src/mds/Locker.cc

```diff
@@ -3634,9 +3634,19 @@ void Locker::_do_snap_update(CInode *in, snapid_t snap, int dirty, snapid_t foll
   if (xattrs) {
     dout(7) << " xattrs v" << i->xattr_version << " -> " << m->head.xattr_version
 	    << " len " << m->xattrbl.length() << dendl;
-    i->xattr_version = m->head.xattr_version;
-    auto p = m->xattrbl.cbegin();
-    decode(*px, p);
+    __u32 total = 0;
+    for (const auto &p : *px)
+      total += (p.first.length() + p.second.length());
+    if ((total + m->xattrbl.length()) >= g_conf()->mds_max_xattr_pairs_size) {
+      dout(10) << "xattr kv pairs size too big. current size: " << total
+             << " new size: " << (total + m->xattrbl.length()) << dendl;
+      // Ignore new xattr (!!!) but increase xattr version
+      i->xattr_version = m->head.xattr_version + 1;
+    } else {
+      i->xattr_version = m->head.xattr_version;
+      auto p = m->xattrbl.cbegin();
+      decode(*px, p);
+    }
```

Comment:
Not sure whether will this cause issues. Because in the client side it succeeds but in MDS side it just discard the xattr.

Maybe we should do:

1, In MDS side to make sure the `mds_max_xattr_pairs_size` option is not allowed to be configured during `runtime`:

```
  69 - name: mds_max_xattr_pairs_size
  70   type: size
  71   level: advanced
  72   desc: maximum aggregate size of extended attributes on a file
  73   default: 64_K
  74   services:
  75   - mds
  76   with_legacy: true
```

Yeah, it doesn't have the `runtime` flags explicitly, but from https://github.com/ceph/ceph/blob/master/src/common/options.h#L387 if there has no any flags set it will allow by default. This should be fixed by adding a flag `FLAG_STARTUP` or something else.

2, Tell the clients what's the max limit of the size allowed when opening the sessions.

3, The clients should check the value size when setting the xattr.

Reply:
> Not sure whether will this cause issues. Because in the client side it succeeds but in MDS side it just discard the xattr.

Yeah, I'm not happy with this either.  The only issue I saw was the client still showing the xattrs that were discarded by the MDS.

Your suggestion sounds good to me for well-behaved clients, i.e. clients which would respect this limit.  What about old clients?  They still need to be handled.

Also, how would you suggest to notify the clients about the limit?  Through the session MDS map?  (Which is the only map being decoded by the kernel client AFAIR).

Reply:
Also, how can we force the client to drop all caps it currently owns at this point so that he becomes aware immediately that the xattrs are not set?  And maybe we should also prevent this client from getting Xw caps...?

Reply:
> > Not sure whether will this cause issues. Because in the client side it succeeds but in MDS side it just discard the xattr.
> 
> Yeah, I'm not happy with this either. The only issue I saw was the client still showing the xattrs that were discarded by the MDS.
> 
> Your suggestion sounds good to me for well-behaved clients, i.e. clients which would respect this limit. What about old clients? They still need to be handled.

I don't think we can gracefully fix it and to work well for the old mds/clients. IMO we can do:

1), in client side, you can hard code and limit it to 64K as default if the MDS doesn't support the new feature. Or we should use the limit passed from MDS. And at the `mds.yaml` just add one warning messages in the option's descrption to tell users it's not safe to decrease it if they are using the old clients.

2), In MDS side, just truncate xattr values to the limitation allowed if the old clients passed large values.


> 
> Also, how would you suggest to notify the clients about the limit? Through the session MDS map? (Which is the only map being decoded by the kernel client AFAIR).

No, just add one member in `MClientSession` and pass it to clients when opening the sessions.

Reply:
> Also, how can we force the client to drop all caps it currently owns at this point so that he becomes aware immediately that the xattrs are not set? And maybe we should also prevent this client from getting Xw caps...?

Yeah, it's a good idea to disallow the old clients to get the `Xx` caps.

Reply:
Silently truncating xattrs is a bit off-putting. We really do not have a nice way to deal with old clients, do we? Disallowing Xx is probably what we would want to do.

Reply:
Ok, I've another version that does what @lxbsz suggested.  But I'm still trying to figure out what's the best way to prevent old clients from getting the Xx caps.  Were you thinking about yet another feature bit for that?

(I've done a quick attempt at adding a feature bit, but looks like the kernel client needs some more changes in order to encode extra bits; the black magic being used in `encode_supported_features()` doesn't seem to support more bits.)

Reply:
Yeah, adding one new feature bit looks good.